### DISCUSSION TITLE
Develop an adaptive test with jsPsych

### DISCUSSION THREAD

**Discussion by FayDu716** [jspsych/jsPsych#2944] [Q&A]
Hi everyone!

I wonder if we can use jsPsych to develop an adaptive test. I want to develop an adaptive test with functions like jumping to next test item when participants provide a correct answer, or jumping to next training item when participants provide an incorrect answer. And in the training item, gradual feedback will be provided to the participant until they select the correct answer.

I don't have any code yet, just wondering if this can be achieved on jsPsych. 
Hoping someone can come up with some suggestion or a solution.

Many thanks,
Faye

**Comment by jodeleeuw** [comment]
Hi @FayDu716,

Yes, you could do this with jsPsych. It'll be a little more complicated than the typical kinds of experiments that we have demos for in the documentation. You would probably need to make a lot of use of [looping timelines](https://www.jspsych.org/7.3/overview/timeline/#looping-timelines) and [dynamic parameters](https://www.jspsych.org/7.3/overview/dynamic-parameters/).

**Reply by FayDu716** [reply]
Hi Josh, Thanks for your response. I have been working on the code lately. Can I post a follow up question?
I want to provide feedback based on the text that the participants write on the survey. 
Here is my code:

> 

`var Item4 = {
      type: jsPsychSurveyHtmlForm,
      preamble: `
      <img src="img/Item4.png" alt="Item4" style="float:center;width:1100px;"></img>
     `,
     html: '<span style = "font-size: 20px">I pick  <input name = "Item4_response1" id = "test-resp-box" size:"50"/>.  </span>',
     autofocus:'test-resp-box',
     button_label: ["submit"],
     data: {
      Item_type: 'Training_Item4',
      Attempt: '1',
      correct_response: 'E'
     },
     on_finish: function(data){
      if (jsPsych.pluginAPI.compareKeys(data.response.Item4_response1, "E")) {
        data.correct = true;
      } else {
        data.correct = false;
      }
      console.log (data.response,data.correct_response);
     }
     };

     var Item4_feedback0= {
      type: jsPsychHtmlButtonResponse,
      stimulus: function (){
       var last_trial_correct = jsPsych.data.get().last(1).values()[0].correct;
       if (last_trial_correct){
        return '<span style = "font-size: 30px" color = "green">Correct!</span>';
        } else {
          return '<span style = "font-size: 30px" color = "red">Incorrect! Try again.</span>';
        }
       },
      choices: ['ok'],
      };

      var Item4_feedback1a = {
      type: jsPsychHtmlButtonResponse,
      stimulus: function(){
       var last_trial_correct = jsPsych.data.get().last(1).values()[0].correct;
       if (last_trial_correct){
        return '<span style = "font-size: 30px" color = "green">Correct!</span>';
        } else {
          return `
          <p style = "font-size: 30px;" "color:red;">Incorrect!</p>
          <p style = "font-size: 30px;" "color:black;">Please think of these questions:</p>
          <p style = "font-size: 30px;" "color:black;">some prompts</p>
          <p style = "font-size: 30px;" "color:black;">Please try again.</p>;
        }
       },
      choices: ['ok'],
      };

      var Item4_feedback1b = {
      type: jsPsychHtmlButtonResponse,
      stimulus: function(){
        var last_trial_correct = jsPsych.data.get().last(1).values()[0].correct;
       if (last_trial_correct){
        return '<span style = "font-size: 30px" color = "green">Correct！</span>';
        } else {
          return `
          <p style = "font-size: 30px;" "color:red;">Incorrect!</p>
          <p style = "font-size: 30px;" "color:black;">Please think of these questions:</p>
          <p style = "font-size: 30px;" "color:black;">Some other prompts</p>
          <p style = "font-size: 30px;" "color:black;">Please try again.</p>
            `;
        }
       },
      choices: ['ok'],
      };

    var Item4_loop1 = {
      timeline: [Item4, Item4_feedback1a],
      loop_function: function(data){
        console.log(jsPsych.data.get().last(2).values()[0]);
        var last_trial_correct = jsPsych.data.get().last(2).values()[0].correct;
        if(last_trial_correct){
          return false;
        } else {
          return true;
        }
      }
    };


    var Item4_node ={
      timeline: [Item4, Item4_feedback0, Item4_loop1],
      conditional_function: function (){
        var data = jsPsych.data.get().last(1).values()[0];
        if(jsPsych.pluginAPI.compareKeys(data.response.Item1_response1, "C")){
          return false;
        } else {
          return true;
        }
      }
    };`
I can use these codes to jump to different feedback based on how many times the participant has tried Item4. But how can I adjust it to jump to different feedback based on the text entered by the participant? For instance, if the participant enters "A", then feedback1a is provided, and if the participant enters "B", then feedback 1b is provided.

Many thanks!
